jaxley/integrate.py

```diff
@@ -69,17 +72,33 @@ def integrate(
     if "i" in module.externals.keys() or data_stimuli is not None:
         if "i" in module.externals.keys():
             if data_stimuli is not None:
-                externals["i"] = jnp.concatenate([externals["i"], data_stimuli[0]])
+                externals["i"] = jnp.concatenate([externals["i"], data_stimuli[1]])
                 external_inds["i"] = jnp.concatenate(
-                    [external_inds["i"], data_stimuli[1].comp_index.to_numpy()]
+                    [external_inds["i"], data_stimuli[2].comp_index.to_numpy()]
                 )
         else:
-            externals["i"] = data_stimuli[0]
-            external_inds["i"] = data_stimuli[1].comp_index.to_numpy()
-    else:
+            externals["i"] = data_stimuli[1]
+            external_inds["i"] = data_stimuli[2].comp_index.to_numpy()
+    elif data_clamps is None:
         externals["i"] = jnp.asarray([[]]).astype("float")
         external_inds["i"] = jnp.asarray([]).astype("int32")
 
+    # If a clamp is inserted, add it to the external inputs.
+    if data_clamps is not None:
+        state_name, clamps, inds = data_clamps
+        if state_name in module.externals.keys():
+            externals[state_name] = jnp.concatenate([externals[state_name], clamps])
+            external_inds[state_name] = jnp.concatenate(
+                [external_inds[state_name], inds.comp_index.to_numpy()]
+            )
+        else:
+            externals[state_name] = clamps
+            external_inds[state_name] = inds.comp_index.to_numpy()
+        # Need to inject a null current for integration if there isn't one
```

Comment:
Removing this actually seems to cause me some problems now, specifically when trying to jit a simulation without data_stimuli, only data_clamps. There's not a problem without jitting and also not a problem when there are both data_stimuli and data_clamps, but leaving out the stimulus causes problems for jitting. 
I will investigate this more and try to come up with a solution, but any insight is welcome of course!

Reply:
Update: the problem seems to be with jitting the grad calculation on a simulation with only data_clamps which results in a segmentation fault. Interestingly, there is no problem jitting the grad calc on a simulation with both data_stimuli and data_clamps, but without the data_stimulate one gets a segfault. Also to note, one can jit a simulation with just data_clamps and one can grad an unjitted simulation with only data clamps, but the combo crashes the kernel.

Reply:
Some minimal code to reproduce the segmentation fault @huangziwei 
```
from jax import config
config.update("jax_enable_x64", True)
config.update("jax_platform_name", "cpu")

import faulthandler
faulthandler.enable()

import jax.numpy as jnp
from jax import jit, value_and_grad, grad

import jaxley as jx
from jaxley_mech.channels.kwu09 import Leak
from jaxley_mech.channels.chen24 import Phototransduction


comp = jx.Compartment()
branch = jx.Branch(comp, nseg=1)
PR = jx.Cell([branch], parents=[-1])

PR.insert(Phototransduction())
PR.insert(Leak())

PR.delete_recordings()
PR.record()

dt = 0.025 # ms
total_time = 10 # ms
t_steps = int(total_time/dt)

def simulate(params):
    R_clamp = jnp.zeros(t_steps) + params[0]["R"]
    data_clamps = PR.branch(0).comp(0).data_clamp("Phototransduction_Stim", R_clamp, data_clamps=None)

    v = jx.integrate(
        PR, 
        delta_t=dt, 
        data_clamps=data_clamps,
        solver="bwd_euler"
        )
    return v[0, :-1]

def loss_fn(params, data):
    sim_soln = simulate(params)
    loss = jnp.sum(jnp.square(sim_soln - data)) / sim_soln.shape[0]
    return loss

grad_fn = jit(grad(loss_fn, argnums=0))

params = [{"R": 1000.0}]

data = jnp.zeros(t_steps) - 30.0

for i in range(10):
    g = grad_fn(params, data)
    print(f"it {i}, g: {g}")

```

Reply:
I think `t_max` is required for `jx.intergrate()` when there're no stimuli / data_stimuli.

Reply:
You're totally right @huangziwei! And the way it was happening was that t_max was triggering the creation of an external null current, and then a lot of calculations were happening that don't actually need to be happening, so I revamped things a bit. I would recommend now looking at the files alone and not the diff view bc that's kind of a headache right now. But I fixed it so that externals can be empty and everything still seems to work, also with not as many function calls and so on. Lmk what you guys think now when you have a chance :)

Reply:
Yes, but now t_max doesn't trigger the creation of an external input